qtconsole/console_widget.py

```diff
@@ -2203,7 +2220,13 @@ def _insert_plain_text(self, cursor, text, flush=False):
                                 QtGui.QTextCursor.MoveAnchor)
 
                     elif act.action == 'newline':
-                        cursor.movePosition(QtGui.QTextCursor.EndOfLine)
+                        self.log.debug(self._prompt)
+                        if cursor.block() != cursor.document().lastBlock():
+                            cursor.movePosition(QtGui.QTextCursor.NextBlock)
+                        else:
+                            cursor.movePosition(QtGui.QTextCursor.EndOfLine,
+                                                QtGui.QTextCursor.MoveAnchor)
+                            cursor.insertText('\n')
```

Comment:
@jsbautista I think what is missing here (and the cause of the failed test) is to not only check if we are over the end of the document but also that the document doesn't end with `self._prompt`. So maybe doing something like:

```suggestion
                        if (
                            cursor.block() != cursor.document().lastBlock()
                            and not cursor.document()
                            .toPlainText()
                            .endswith(self._prompt)
                        ):
                            cursor.movePosition(QtGui.QTextCursor.NextBlock)
                        else:
                            cursor.movePosition(
                                QtGui.QTextCursor.EndOfLine,
                                QtGui.QTextCursor.MoveAnchor,
                            )
                            cursor.insertText("\n")
```

Could help tests pass